In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchsummary import summary
from tqdm import tqdm

In [2]:
from model2 import Net
from train_test_loop import *
from data_utils import *

In [3]:
torch.manual_seed(1)

In [4]:
BATCH_SIZE=128
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

In [5]:
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
       BatchNorm2d-3            [-1, 8, 26, 26]              16
           Dropout-4            [-1, 8, 26, 26]               0
            Conv2d-5           [-1, 16, 24, 24]           1,152
              ReLU-6           [-1, 16, 24, 24]               0
       BatchNorm2d-7           [-1, 16, 24, 24]              32
           Dropout-8           [-1, 16, 24, 24]               0
            Conv2d-9           [-1, 10, 24, 24]             160
        MaxPool2d-10           [-1, 10, 12, 12]               0
           Conv2d-11           [-1, 16, 10, 10]           1,440
             ReLU-12           [-1, 16, 10, 10]               0
      BatchNorm2d-13           [-1, 16, 10, 10]              32
          Dropout-14           [-1, 16,

In [6]:
default_train_transforms = transforms.Compose([
                                        transforms.RandomRotation((-7.0, 7.0), fill=(1,)),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))]) #None
default_test_transforms = transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ]) #None

In [7]:
train_dataloader, test_dataloader = get_augmented_MNIST_dataset("~/work/data/", 
                                    train_tfms=default_train_transforms, test_tfms=default_test_transforms,
                                    batch_sz=BATCH_SIZE)

In [10]:
optimizer = torch.optim.SGD(model.parameters(), lr=0.05, momentum=0.9)
scheduler = get_schedulder(optimizer, "StepLR", step_size=6, gamma=0.25)

In [11]:
EPOCHS=15
for i in range(EPOCHS):
    print("Running EPOCH: "+str(i+1))
    train(model, device, train_dataloader, optimizer=optimizer, epoch=i)
    test(model, device, test_dataloader)

Running EPOCH: 1


Loss=0.06363967061042786 Batch_id=468 Accuracy=98.08: 100%|██████████| 469/469 [00:01<00:00, 306.88it/s] 



Test set: Average loss: 0.0353, Accuracy: 9885/10000 (98.85%)

Running EPOCH: 2


Loss=0.061006009578704834 Batch_id=468 Accuracy=98.41: 100%|██████████| 469/469 [00:01<00:00, 289.29it/s]



Test set: Average loss: 0.0301, Accuracy: 9914/10000 (99.14%)

Running EPOCH: 3


Loss=0.04641694948077202 Batch_id=468 Accuracy=98.63: 100%|██████████| 469/469 [00:01<00:00, 300.91it/s] 



Test set: Average loss: 0.0250, Accuracy: 9918/10000 (99.18%)

Running EPOCH: 4


Loss=0.02934526838362217 Batch_id=468 Accuracy=98.68: 100%|██████████| 469/469 [00:01<00:00, 303.89it/s]  



Test set: Average loss: 0.0273, Accuracy: 9912/10000 (99.12%)

Running EPOCH: 5


Loss=0.033484023064374924 Batch_id=468 Accuracy=98.70: 100%|██████████| 469/469 [00:01<00:00, 305.73it/s]



Test set: Average loss: 0.0264, Accuracy: 9921/10000 (99.21%)

Running EPOCH: 6


Loss=0.0482063852250576 Batch_id=468 Accuracy=98.79: 100%|██████████| 469/469 [00:01<00:00, 288.77it/s]   



Test set: Average loss: 0.0280, Accuracy: 9914/10000 (99.14%)

Running EPOCH: 7


Loss=0.06527023017406464 Batch_id=468 Accuracy=98.79: 100%|██████████| 469/469 [00:01<00:00, 284.63it/s]  



Test set: Average loss: 0.0277, Accuracy: 9912/10000 (99.12%)

Running EPOCH: 8


Loss=0.03801874443888664 Batch_id=468 Accuracy=98.87: 100%|██████████| 469/469 [00:01<00:00, 277.89it/s]  



Test set: Average loss: 0.0292, Accuracy: 9916/10000 (99.16%)

Running EPOCH: 9


Loss=0.01813000626862049 Batch_id=468 Accuracy=98.87: 100%|██████████| 469/469 [00:01<00:00, 294.30it/s]  



Test set: Average loss: 0.0232, Accuracy: 9925/10000 (99.25%)

Running EPOCH: 10


Loss=0.005274960771203041 Batch_id=468 Accuracy=98.89: 100%|██████████| 469/469 [00:01<00:00, 296.52it/s] 



Test set: Average loss: 0.0224, Accuracy: 9930/10000 (99.30%)

Running EPOCH: 11


Loss=0.027188187465071678 Batch_id=468 Accuracy=99.00: 100%|██████████| 469/469 [00:01<00:00, 305.59it/s] 



Test set: Average loss: 0.0230, Accuracy: 9935/10000 (99.35%)

Running EPOCH: 12


Loss=0.023167172446846962 Batch_id=468 Accuracy=98.99: 100%|██████████| 469/469 [00:01<00:00, 304.64it/s] 



Test set: Average loss: 0.0224, Accuracy: 9930/10000 (99.30%)

Running EPOCH: 13


Loss=0.0857502818107605 Batch_id=468 Accuracy=98.98: 100%|██████████| 469/469 [00:01<00:00, 289.18it/s]   



Test set: Average loss: 0.0228, Accuracy: 9944/10000 (99.44%)

Running EPOCH: 14


Loss=0.012845419347286224 Batch_id=468 Accuracy=98.99: 100%|██████████| 469/469 [00:01<00:00, 298.63it/s] 



Test set: Average loss: 0.0264, Accuracy: 9916/10000 (99.16%)

Running EPOCH: 15


Loss=0.053014159202575684 Batch_id=468 Accuracy=98.98: 100%|██████████| 469/469 [00:01<00:00, 296.70it/s] 



Test set: Average loss: 0.0199, Accuracy: 9944/10000 (99.44%)



#### Target: 
Reduce the model size to less than 8k params (primary), and tweak with lr/scheduler

#### Result: 
Accuracy 99.444 but fluctuating. Model size: 7736

#### Analysis: 

Now trying to get the model size down first. The change is in the backbone CNN which goes like this: input block -> 2 conv blocks (conv + ReLU) -> transition block (2x2 maxpool) -> 4 conv blocks (conv + ReLU) -> final conv2d layer. The difference from Step 1 is the number of channels: Step1 has a channel progression like 1 -> 32 -> 64 -> 128 -> 32 -> 64 -> 128 -> 10 -> 10, where as we don't cross 16 channels in any conv2d layer here.  The new channel progression is 1 -> 8 -> 16 -> 10 -> 16 -> 16 -> 10 with a GAP of kernel size 6 before the last layer. CHange the lr scheduler with lr = 0.05, step_size = 6 and gamma = 0.25. The final change is the removal of one conv block just before the GAP layer, as an experiment to reduce model size, and the number of params is 7,736 (within our range). I selected that particular conv block since the number of input and output channels are the same, so I thought I could remove since number of input and output features is the same. (May work for smaller datasets like MNIST, but may fail on larger datasets since many more features are present in larger datasets). The test accuracy for 15 epochs: 99.44% but is fluctuating. While we do achieve the required test accuracy, it is not consistent, so may be we are reducing the lr too fast (_i.e._ we should increase step_size or decrease gamma or increase lr or a combination of these three).